### ISSUE TITLE
1.1.1f： If call SSL_key_update while writes are pending（no-blocking socket），next SSL_write or SSL_read will return fail

### ISSUE THREAD

**Issue by yanchangjia** [openssl/openssl#12485]
In TLS servers where asynchronous reads and writes may be flowing concurrently(non-block socket) , OpenSSL's TLS 1.3 can't support call SSL_key_update between SSL_ERROR_WANT_WRITE return by SSL_write and next SSL_write or SSL_read.
it will fail with the following error:
SSLfatal(s, SSL_AD_INTERNAL_ERROR, SSL_F_SSL3_WRITE_BYTES,SSL_R_BAD_LENGTH);

I have seen similar cases #8677, but our case is that call SSL_key_update while writes are pending（no-blocking socket).
may be our using logic is wrong?
Is there any documentation that supports this issue?

this is test case：
```
ssl = SSL_new(Context);		
new_sd = create_socket();	
if (ioctl(iSd, FIONBIO, &nonBlock) == SOCKET_ERROR)
{
	printf("ioctl() failed with error %d\n", ipsi_get_last_socket_error());
	return;
} 
ret =SSL_setSockId(ssl, (SEC_SOCK)iSd);
while(1){
		err = SSL_accept(ssl);  	//connect to the server
		if (err == 1)
		{
                    break;
		}
}
// step1 :Write a large message on the server，until SSL_write return -1 with SSL_ERROR_WANT_WRITE
	while(1)
	{
		ret = SSL_write(ssl, buf_write, sizeof(buf_write));
		if(ret == -1 && SSL_get_error(ssl, ret) == SSL_ERROR_WANT_WRITE)
		{
			printf("SSL_ERROR_WANT_WRITE exit\n");
			break;
		}
	}	  
//step2: server call SSL_key_update
	ret = SSL_key_update(ssl, SSL_KEY_UPDATE_NOT_REQUESTED);  //SSL_KEY_UPDATE_REQUESTED
	TEST_ASSERT(ret == TEST_SSL_SUCCESS,"SSL_key_update");

//step3: do SSL_write again,it will fail and return error :SSL_R_BAD_LENGTH
	ret = SSL_write(ssl， buf_write, sizeof(buf_write));
	TEST_ASSERT(ret != -1,"SSL_write fail");
        err = SSL_get_error(ssl, ret);
       printf("last write err=%d\n",err);
       printf("\nret: %d, SSL_getLastError: %d\n",ret, ERR_GET_REASON(ERR_peek_error()));  //**SSL_R_BAD_LENGTH**

```

**Comment by kaduk** [comment]
Isn't this a duplicate of #12411 ?

**Comment by yanchangjia** [comment]
> Isn't this a duplicate of #12411 ?

YES，issue 12411 no one responded. I cannot guarantee that this is a bug or a usage problem。 so I submit a issue:question

**Comment by kaduk** [comment]
It's probably going to be either me or @mattcaswell responding -- Matt was on holiday last week and my attention was focused elsewhere by a big deadline.  I intend to look at it "soon" but cannot commit to having the time available quite yet.

**Comment by yanchangjia** [comment]
@kaduk @mattcaswell 
Can you pay attention to this issue?

**Comment by kaduk** [comment]
It turns out that this is a fair bit easier to understand from the `SSL_key_update()` side than the state-machine side.
Namely, we only check `!SSL_is_init_finished()` before forcibly setting "in init" via `ossl_statem_set_in_init()`.
Presumably we need to be adding some kind of check against `rlayer.wpend_tot` to `SSL_key_update()` (the simplest thing would be to just deny the request to do a key update if writes are still pending, though if that was too disruptive we could add another flag bit and check it later).
@mattcaswell does the simple approach (return an error from `SSL_key_update()` if writes are pending) seem okay to you?

**Comment by mattcaswell** [comment]
> @mattcaswell does the simple approach (return an error from SSL_key_update() if writes are pending) seem okay to you?

This seems like a reasonable approach. We better just be sure to update the documentation to make sure we capture this restriction.

**Comment by yanchangjia** [comment]
@kaduk @mattcaswell 
You confirm that this is not an interface usage logic problem.
Have you confirmed that this is a bug now? Do you plan to implement changes in later versions?

**Comment by mattcaswell** [comment]
> You confirm that this is not an interface usage logic problem.
Have you confirmed that this is a bug now? Do you plan to implement changes in later versions?

Yes, it looks like a bug. I'm hoping @kaduk is going to come up with a fix :-)

**Comment by kaduk** [comment]
I have a todo item, yes.  This week is busy for me, though, so no ETA

**Comment by kaduk** [comment]
> You confirm that this is not an interface usage logic problem.
Have you confirmed that this is a bug now? 

I'd follow-up on Matt's answer to say that this actually is an application's interface-usage logic problem, but that there's a bug in that the library should return an error gracefully in this situation and not corrupt the state of the SSL object.  The application will still need to change its code to not attempt to call `SSL_key_update()` while writes are pending.

**Comment by yangyangtiantianlonglong** [comment]
@kaduk @mattcaswell   
When SSL_key_update() is called during non-blocking read, the same problem can appear？

**Comment by kaduk** [comment]
I expect so, yes.
The sketch of a patch I have in mind will cover that case, too.

**Comment by kaduk** [comment]
Having looked at this a bit, it may actually be pretty easy to get the more-preferred functionality where we allow the API call at any time and defer the write until it is safe to do so.  The code change itself might look like https://github.com/kaduk/openssl/commit/b653b50aca4fd6e75048b9929b2d067849343aaa but it needs tests and maybe docs before it's ready to become a PR.  It might be useful to have specific tests for when a local keyupdate request is triggered but a peer's update request comes in before the locally-triggered one is written.

**Comment by yangyangtiantianlonglong** [comment]
https://github.com/openssl/openssl/blob/3094351625f0b222f92c22ce4943461df8c7e301/ssl/s3_lib.c#L4582-L4604
@kaduk 
1. The preceding modification only does not update the key when the Complete record is not written. However, the same problem occurs when the Complete record is not read. Therefore, check like  lines 4587 and 4588.

2. The key status is successfully updated, but the key update is not triggered. I think the renegotiation check is more reasonable. The new interface (```ssl_keyupdate_check```) helps users understand why the key is not updated.

**Comment by kaduk** [comment]
TLS 1.2 renegotiation has different semantics than the TLS 1.3 KeyUpdate -- renegotiation is in some sense a synchronization point at which both read and write keys are changed, whereas the KeyUpdate is a unilateral "I am changing my keys now" message.  You can request the other party to do the same, but that happens asynchronously.  So I don't think that KeyUpdate needs to depend on having read a complete record.

@yangyangtiantianlonglong  I don't understand what you mean in your (2) -- which "key status" is updated, and when?  What is the "key update" that is not triggered?  I think I understand that you are proposing a new (perhaps `static_inline`) API called `ssl_keyupdate_check()` to encapsulate the needed logic for whether it is safe to do a key update ... but do you mean just the TLS 1.3 KeyUpdate or is it supposed to be of broader applicability?

**Comment by yangyangtiantianlonglong** [comment]
https://github.com/kaduk/openssl/commit/b653b50aca4fd6e75048b9929b2d067849343aaa
(2)RECORD_LAYER_write_pending  return 1, SSL_key_update return 1, The next step does not send the keyupdate msg.
However, the user believes that the update has been successful；SSL_key_update is successful.

**Comment by t8m** [comment]
Reopening as for 1.1.1 there needs to be a separate PR.

**Comment by link2xt** [comment]
@t8m 
Isn't it fixed by #16098 ?

**Comment by t8m** [comment]
Yes, it is. Closing.
